sys/net/gnrc/transport_layer/tcp/gnrc_tcp_eventloop.c

```diff
@@ -0,0 +1,251 @@
+    uint16_t ctl = 0;
+    uint16_t src = 0;
+    uint16_t dst = 0;
+    uint8_t hdr_size = 0;
+    uint8_t syn = 0;
+    gnrc_pktsnip_t *ip = NULL;
+    gnrc_pktsnip_t *reset = NULL;
+    gnrc_tcp_tcb_t *tcb = NULL;
+    tcp_hdr_t *hdr;
+
+    /* mark TCP header */
+    gnrc_pktsnip_t *tcp = gnrc_pktbuf_start_write(pkt);
+    if (tcp == NULL) {
+        DEBUG("gnrc_tcp_eventloop : _receive() : can't write to packet\n");
+        gnrc_pktbuf_release(pkt);
+        return -EACCES;
+    }
+    pkt = tcp;
+
+    /* Get IP Header */
+    LL_SEARCH_SCALAR(pkt, ip, type, GNRC_NETTYPE_IPV6);
+    assert(ip != NULL);
+
+    /* Get TCP Header */
+    LL_SEARCH_SCALAR(pkt, tcp, type, GNRC_NETTYPE_TCP);
+    assert(tcp != NULL);
+
+    /* Extract control bits */
+    ctl = byteorder_ntohs(((tcp_hdr_t *) tcp->data)->off_ctl);
+
+    /* Validate Offset */
+    if (GET_OFFSET(ctl) < OPTION_OFFSET_BASE) {
+        DEBUG("gnrc_tcp_eventloop : _receive() : unexpected Offset Value\n");
+        gnrc_pktbuf_release(pkt);
+        return -ERANGE;
+    }
+
+    /* Calculate tcp header size */
+    hdr_size = GET_OFFSET(ctl) * 4;
+
+    /* Mark TCP-Header, if it contains any payload */
+    if ((pkt->type == GNRC_NETTYPE_TCP) && (pkt->size != hdr_size)) {
+        tcp = gnrc_pktbuf_mark(pkt, hdr_size, GNRC_NETTYPE_TCP);
+        if (tcp == NULL) {
+            DEBUG("gnrc_tcp_eventloop : _receive() : Header marking failed\n");
+            gnrc_pktbuf_release(pkt);
+            return -ENOMSG;
+        }
+        pkt->type = GNRC_NETTYPE_UNDEF;
+    }
+    hdr = (tcp_hdr_t *)tcp->data;
+
+    /* Validate Checksum */
+    if (byteorder_ntohs(hdr->checksum) != _pkt_calc_csum(tcp, ip, pkt)) {
+        DEBUG("gnrc_tcp_eventloop : _received() : Invalid checksum\n");
+        gnrc_pktbuf_release(pkt);
+        return -EINVAL;
+    }
+
+    /* De-multiplex connection */
+    syn = ((ctl & MSK_SYN_ACK) == MSK_SYN);
+    src = byteorder_ntohs(hdr->src_port);
+    dst = byteorder_ntohs(hdr->dst_port);
+
+    /* Iterate over tcb list to find the connections tcb */
+    mutex_lock(&_list_gnrc_tcp_tcb_lock);
+    tcb = _list_gnrc_tcp_tcb_head;
+    while (tcb) {
+        /* If SYN is set, a connection is listening on that port ... */
+        if (syn && *(tcb->local_port) == dst && tcb->state == LISTEN) {
+            /* ... If this Paket is for the specified address or the local addr is unspec */
+            if (ipv6_addr_equal((ipv6_addr_t *)tcb->local_addr, &((ipv6_hdr_t * ) ip->data)->dst)
+            || ipv6_addr_is_unspecified((ipv6_addr_t *)tcb->local_addr)
+            ) {
+                break;
+            }
+        }
+
+        /* If portnumbers and addresses match and SYN only is not set */
+        /* NOTE: Make addresscheck IP agnostic if IPv6 is ever implemented !!! */
+        if (!syn && *(tcb->local_port) == dst && *(tcb->peer_port) == src
+        && ipv6_addr_equal((ipv6_addr_t *)tcb->peer_addr, &((ipv6_hdr_t *) ip->data)->src)
```

Comment:
I don't see a indentation error and whats wrong with the && here? I do that in the entire module.